sky/packages/sky/lib/src/rendering/stack.dart

```diff
@@ -226,14 +230,112 @@ class RenderStack extends RenderBox with ContainerRenderObjectMixin<RenderBox, S
   }
 
+  void paintStack(PaintingContext context, Offset offset);
+
   void paint(PaintingContext context, Offset offset) {
     if (_hasVisualOverflow) {
       context.canvas.save();
       context.canvas.clipRect(offset & size);
-      defaultPaint(context, offset);
+      paintStack(context, offset);
       context.canvas.restore();
     } else {
-      defaultPaint(context, offset);
+      paintStack(context, offset);
     }
   }
 }
+
+/// Implements the stack layout algorithm
+///
+/// In a stack layout, the children are positioned on top of each other in the
+/// order in which they appear in the child list. First, the non-positioned
+/// children (those with null values for top, right, bottom, and left) are
+/// initially layed out and placed in the upper-left corner of the stack. The
+/// stack is then sized to enclose all of the non-positioned children. If there
+/// are no non-positioned children, the stack becomes as large as possible.
+///
+/// The final location of non-positioned children is determined by the alignment
+/// parameters. The left of each non-positioned child becomes the
+/// difference between the child's width and the stack's width scaled by
+/// horizontalAlignment. The top of each non-positioned child is computed
+/// similarly and scaled by verticalAlignement. So if the alignment parameters
+/// are 0.0 (the default) then the non-positioned children remain in the
+/// upper-left corner. If the alignment parameters are 0.5 then the
+/// non-positioned children are centered within the stack.
+///
+/// Next, the positioned children are laid out. If a child has top and bottom
+/// values that are both non-null, the child is given a fixed height determined
+/// by deflating the width of the stack by the sum of the top and bottom values.
+/// Similarly, if the child has rigth and left values that are both non-null,
+/// the child is given a fixed width. Otherwise, the child is given unbounded
+/// space in the non-fixed dimensions.
+///
+/// Once the child is laid out, the stack positions the child according to the
+/// top, right, bottom, and left offsets. For example, if the top value is 10.0,
+/// the top edge of the child will be placed 10.0 pixels from the top edge of
+/// the stack. If the child extends beyond the bounds of the stack, the stack
+/// will clip the child's painting to the bounds of the stack.
+class RenderStack extends _RenderStackBase {
+  RenderStack({
+    List<RenderBox> children,
+    double horizontalAlignment: 0.0,
+    double verticalAlignment: 0.0
+  }) : super(
+   children: children,
+   horizontalAlignment: horizontalAlignment,
+   verticalAlignment: verticalAlignment
+ );
+
+  void paintStack(PaintingContext context, Offset offset) {
+    defaultPaint(context, offset);
+  }
+}
+
+/// Implements the same layout algorithm as RenderStack but only paints the child
+/// specified by index.
+class RenderIndexedStack extends _RenderStackBase {
```

Comment:
Add a comment saying that this class's algorithms are still O(N) in the length of children even though they're painting only one item, since people might expect this to be O(1).